package/src/cpumon.cpp

```diff
@@ -57,3 +58,56 @@ std::map<std::string, double> const cpumon::get_json_average_stats(
   std::map<std::string, double> empty_average_stats{};
   return empty_average_stats;
 }
+
+// Collect related hardware information
+std::map<std::string, std::map<std::string, std::string>> const cpumon::get_hardware_info() {
+  std::map<std::string, std::map<std::string, std::string>> result{};
+
+  // Read some information from /proc/cpuinfo
+  std::ifstream cpuInfoFile{"/proc/cpuinfo"};
+  if(!cpuInfoFile.is_open()) {
+    std::cerr << "Failed to open /proc/cpuinfo" << std::endl;
+    return result;
+  }
+
+  // Metrics to read from the input
+  std::vector<std::string> metrics{"processor", "model name", "siblings", "cpu cores"};
+  unsigned int nCPU = 0, nSiblings = 0, nCores = 0;
+
+  // Loop over the file
+  std::string line;
+  while (std::getline(cpuInfoFile,line)) {
+    if (line.empty()) continue;
+    size_t splitIdx = line.find(":");
+    std::string val;
+    if (splitIdx != std::string::npos) {
+      val = line.substr(splitIdx + 1);
+      if (val.empty()) continue;
+      for (const auto& metric : metrics) {
+        if (line.size() >= metric.size() && line.compare(0, metric.size(), metric) == 0) {
+          if (metric == "processor") nCPU++;
+          else if (metric == "siblings") { if(nSiblings == 0) nSiblings = std::stoi(val); }
+          else if (metric == "cpu cores") { if(nCores == 0) nCores = std::stoi(val); }
+          else if (result["cpu"][metric].empty()) {
+            result["cpu"][metric] = std::regex_replace(val, std::regex("^\\s+|\\s+$"), "");
+          }
+        } // end of metric check
+      } // end of populating metrics
+    } // end of seperator check
+  } // end of reading cpuInfoFile
+
+  // The clear assumption here is that there is a single type of processor installed
+  // nCPU = nSockets * nSiblings
+  // nSiblings = nCoresPerSocket * nThreadsPerCore
+  unsigned int nSockets = nCPU / nSiblings;
+  unsigned int nThreads = nSiblings / nCores;
+  result["cpu"]["nCPU"] = std::to_string(nCPU);
+  result["cpu"]["nSockets"] = std::to_string(nSockets);
+  result["cpu"]["nCoresPerSocket"] = std::to_string(nCores);
+  result["cpu"]["nThreadsPerCore"] = std::to_string(nThreads);
```

Comment:
I don't like that these values com out as strings, that seems unnatural. So thinking about this maybe the underlying storage object should be a JSON entity, instead of the more rigid map of strings. That would allow for a more natural representation of this information.

If we were re-mapping post-facto then we'd have to somehow communicate the intrinsic type back to the main program, which seems complicated and not necessary if we went the JSON route. As this information is gathered only once then performance isn't a concern.

Reply:
Done (see https://github.com/HSF/prmon/pull/116#issuecomment-638838785 for more details).